### PR TITLE
Wip add smart

### PR DESCRIPTION
**Pull request by GregMeno**
@dmick 
Adds a module to wrap the mess that is smartctl, reports only reallocated_sectors

### CODE REVIEW

File: salt/srv/salt/_modules/smart.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    mira048.front.sepia.ceph.com:
+        Firmware Version   : V1.49 2011-08-10
+    '''
+    firmware = "/usr/sbin/cli64 sys info | grep -i firm | awk '{print $5}' | cut -d '-' -f1"
+
+    p = Popen(firmware, stdout=PIPE, stderr=PIPE, shell=True)
+    stdout, stderr = p.communicate()
+
+    if int(stdout) < 2011:
+        return 1, '', 'Controller needs newer firmware for S.M.A.R.T. support'
```
Comment (jcsp) [inline salt/srv/salt/_modules/smart.py]: 0, 1, 2 appear to be magic numbers, should be defined as named constants somewhere

File: salt/srv/salt/_modules/smart.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    ubuntu@mira106:~$ sudo salt 'mira048*' cmd.run "/usr/sbin/cli64 sys info | grep -i firm "
+    mira048.front.sepia.ceph.com:
+        Firmware Version   : V1.49 2011-08-10
+    '''
+    firmware = "/usr/sbin/cli64 sys info | grep -i firm | awk '{print $5}' | cut -d '-' -f1"
+
+    p = Popen(firmware, stdout=PIPE, stderr=PIPE, shell=True)
+    stdout, stderr = p.communicate()
+
+    if int(stdout) < 2011:
```
Comment (jcsp) [inline salt/srv/salt/_modules/smart.py]: You need to handle the exception if stdout happens not to be an int

File: salt/srv/salt/_modules/smart.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+def _check_firmware_areca():
+    '''
+    ubuntu@mira106:~$ sudo salt 'mira048*' cmd.run "/usr/sbin/cli64 sys info | grep -i firm "
+    mira048.front.sepia.ceph.com:
+        Firmware Version   : V1.49 2011-08-10
+    '''
+    firmware = "/usr/sbin/cli64 sys info | grep -i firm | awk '{print $5}' | cut -d '-' -f1"
+
+    p = Popen(firmware, stdout=PIPE, stderr=PIPE, shell=True)
```
Comment (jcsp) [inline salt/srv/salt/_modules/smart.py]: You need to handle the error code if this gives you a nonzero return value

File: salt/srv/salt/_modules/smart.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+log = logging.getLogger(__name__)
+__opts__ = salt.config.minion_config('/etc/salt/minion')
+__grains__ = salt.loader.grains(__opts__)
+
+ARECA = 'areca'
+MDADM = 'mdadm'
+
+
+def _get_raid_devices():
```
Comment (jcsp) [inline salt/srv/salt/_modules/smart.py]: dangerously misleading method name -- you're piping lspci into "head -1" so arbitrarily picking the first raid controller (sorted by PCI bus ID)

File: salt/srv/salt/_modules/smart.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+
+def get_errors():
+    '''
+    Iterate all drives we can discover and return a dictionary device: specific SMART failure states
+    '''
+    # areca hardware raid
+    drives = 0
+    report = {}
+    _, pci, _ = _get_raid_devices()
```
Comment (jcsp) [inline salt/srv/salt/_modules/smart.py]: First item in tuple returned from _get_raid_devices() is an error code, but you assigning it to `_` and ignoring it here

File: salt/srv/salt/_modules/smart.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    '''
+    invokes smartctl using the appropiate controller
+    '''
+    scsi_dev = scsi_dev.strip()
+    if raid_type == ARECA:
+        fullcommand = ['smartctl', '-a', '-d', 'areca,%s' % drive, scsi_dev]
+    elif raid_type == MDADM:
+        fullcommand = ['smartctl', '-a', '-d', 'ata', '/dev/%s' % drive]
+    elif raid_type is None:
+        fullcommand = ['smartctl', '-a', '-d', 'sat', '/dev/%s' % drive]
```
Comment (jcsp) [inline salt/srv/salt/_modules/smart.py]: You need an else: here in case raid_type is something unexpected

**Comment by GregMeno** [comment]
@jcsp Thank you for the feedback!  I will clean this up.

**Pull request merged**